design/design-generic-references.md

```diff
@@ -0,0 +1,211 @@
+# Generic Resource References
+
+* Owner: Muvaffak Onuş (@muvaf)
+* Reviewers: Crossplane Maintainers
+* Status: Accepted
+
+## Background
+
+In many cases, Crossplane managed resources require information that is not
+available when the user creates the resource. One example is that if you'd like to create
+a `Subnet`, you need to supply a VPC ID but if you create the `VPC` at the same
+time then the ID will be available only after the creation, so you need to extract
+the ID and use it in `Subnet`. Today, it's possible to let the provider do this
+operation using cross-resource references. However, this requires knowing the kind of
+the referenced resource and how to extract the value during development time of
+the provider. There are several cases where this is not possible:
+
+* The referencer field may not always target a single kind. For example, a URL
+  field in Route53 could get populated by a URL of an S3 bucket or access endpoint
+  of an RDS Instance.
+* Instead of hard-coding region for all managed resources, you may want to refer
+  to a `ConfigMap` or a `Secret` and let the managed resource take that
+  information from there.
+* In Composition, you can use values of the fields only between the children of
+  the same composite instance, but there are cases where you'd like to use a
+  separate resource that is not bound to any composite instance to store the
+  information and serve all composites.
+
+In order to enable such use cases, we need a generic version of the cross-resource
+references where you can specify all metadata required to find the referenced
+resource as well as details about how to extract the value you need and where to
+put it in the referencer object.
+
+## Proposal
+
+All managed resources will have a top-level spec field that lists the generic
+references. They will be resolved in order and failure of any will cause
+reconciliation to fail. Inspired from the [latest developments in API conventions](https://github.com/kubernetes/community/pull/5748)
+regarding multi-kind references, the API will look like the following:
+```yaml
+# A generic reference to another managed resource.
+spec:
+  externalValues:
+  - fromObject:
+      group: ec2.aws.crossplane.io
+      version: v1alpha1
+      resource: vpcs
+      name: main-vpc
+      fieldPath: spec.forProvider.cidrBlock
+    toFieldPath: spec.cidrBlock
+```
+```yaml
+# A generic reference to a namespaced Kubernetes resource.
+spec:
+  externalValues:
+  - fromObject:
```

Comment:
Currently we support both explicit references and selectors. This enables scenarios like "use the subnets created by such and such XR, even though I don't know their Kubernetes names ahead of time". I'd prefer this generic reference approach to have similar functionality (especially if we deprecate the existing approach).

Reply:
Good idea! Added `matchLabels` and made `name` optional. Had a bit of back and forth with having `ref`/`selector` under `fromObject` but feels like all fields under `fromObject` are in the same class of reducing the set of possibilities for the selection hence put them directly there but not feeling strongly about either way.

Reply:
@morningspace just added more details regarding the selector.